### PR TITLE
[#4012] improvement(client-python): Refactor Error Handling in client-python

### PR DESCRIPTION
**Pull request by noidname01**
### What changes were proposed in this pull request?

* Refactor the error handling structure, each API can implement their own error handler to raise custom exceptions
* Add unit test for error handler, but unit tests and integration tests for each API(ex. metalake, catalog, schema) have not been added, I will create issues for them.
	- [ ] Add Metalake Error Handler and related exceptions, test cases
	- [ ] Add Catalog Error Handler and related exceptions, test cases
	- [ ] Add Schema Error Handler and related exceptions, test cases
	- [ ] Add OAuth Error Handler and related exceptions, test cases
* Create `gravitino/exceptions/base.py` to define all the exceptions.
* Remove some unused files and exceptions

### Why are the changes needed?

Fix: #4012

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT added and one IT added, test with `./gradlew clients:client-python:test`

### CODE REVIEW

File: clients/client-python/gravitino/client/gravitino_client_base.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def load_metalake(self, name: str) -> GravitinoMetalake:
 
         self.check_metalake_name(name)
         response = self._rest_client.get(
-            GravitinoClientBase.API_METALAKES_IDENTIFIER_PATH + name
+            GravitinoClientBase.API_METALAKES_IDENTIFIER_PATH + name,
```
Comment (jerryshao) [inline clients/client-python/gravitino/client/gravitino_client_base.py:78]: What is the purpose of this change?

File: clients/client-python/gravitino/dto/responses/base_response.py [hunk 2]
```diff
@@ -35,7 +35,6 @@ def code(self) -> int:
 
     def validate(self):
         """Validates the response code.
-        TODO: @throws IllegalArgumentException if code value is negative.
+        @throws IllegalArgumentException if code value is negative.
         """
-        if self._code < 0:
-            raise ValueError("code must be >= 0")
+        assert self._code >= 0, "code must be >= 0"
```
Comment (jerryshao) [inline clients/client-python/gravitino/dto/responses/base_response.py:40]: Will it throw an `IllegalArgumentException` here?

Comment (noidname01) [reply clients/client-python/gravitino/dto/responses/base_response.py:40]: No, because `assert` can only throw `AssertError`, it can't throw custom exceptions. I think I will create another issue to fix all `assert` problem, otherwise this PR will include too many changes.

File: clients/client-python/gravitino/client/gravitino_client_base.py (hunk 1 above)
Comment (noidname01) [reply clients/client-python/gravitino/client/gravitino_client_base.py:78]: Oh I add some arguments here when I tested, I will fix it.

File: clients/client-python/gravitino/dto/responses/base_response.py (hunk 2 above)
Comment (jerryshao) [reply clients/client-python/gravitino/dto/responses/base_response.py:40]: Can you please create an issue to track this in case you forget?

Comment (noidname01) [reply clients/client-python/gravitino/dto/responses/base_response.py:40]: Have created issue #4100

File: clients/client-python/gravitino/client/gravitino_client_base.py (hunk 1 above)
Comment (noidname01) [reply clients/client-python/gravitino/client/gravitino_client_base.py]: Fixed.

**Comment by noidname01** [comment]
@ian910297 @rich7420 Can you please help to review this?

**Review by ian910297 — COMMENTED** [review]
It's okay for me
(O) Add error handler
(O) Do not inherit Exception class directly. Use RuntimeError, ValueError
(O) Consolidate all exceptions into a single file for easier use.

File: clients/client-python/gravitino/utils/http_client.py [hunk 3]
```diff
@@ -154,19 +177,38 @@ def _request(
...
+
+        if not isinstance(error_handler, ErrorHandler):
+            raise UnknownError(
+                f"Unknown error handler {type(error_handler).__name__}, error response body: {resp}"
+            ) from None
+
+        error_handler.handle(resp)
+        raise UnknownError(
+            f"Error handler {type(error_handler).__name__} can't handle this response, error response body: {resp}"
+        ) from None
```
Comment (jerryshao) [inline clients/client-python/gravitino/utils/http_client.py:193]: I assume that `error_handler.handle(resp)` will handle all the error cases, how can we reach to this point? From my understanding that code will never reach to here, am I right?

```
        raise UnknownError(
            f"Error handler {type(error_handler).__name__} can't handle this response, error response body: {resp}"
        ) from None
```

**Comment by jerryshao** [comment]
Overall LGTM, just one comment, @noidname01 can you please confirm this?

File: clients/client-python/gravitino/utils/http_client.py (hunk 3 above)
Comment (noidname01) [reply clients/client-python/gravitino/utils/http_client.py:193]: Yeah you're right, I add this because `pylint` doesn't let me to just put the `error_handler.handle(resp)` without return or raise, I think it can't analyze that the handler would raise exception anyway. I should add comment for it.

But on the other hand, maybe one error handler don't want to define the default behavior, this can be the final line of defense.

Comment (jerryshao) [reply clients/client-python/gravitino/utils/http_client.py:193]: Can you please add a comment on this code?

Comment (noidname01) [reply clients/client-python/gravitino/utils/http_client.py:196]: Added!

**Review by jerryshao — APPROVED** [review]

**Comment by jerryshao** [comment]
Thanks @noidname01 for your work, can you please continue on the left works, thanks.
